src/Microsoft.AspNetCore.Mvc.Testing/WebApplicationTestFixture.cs

```diff
@@ -12,20 +14,30 @@ namespace Microsoft.AspNetCore.Mvc.Testing
     /// <summary>
     /// Fixture for bootstrapping an application in memory for functional end to end tests.
     /// </summary>
+    /// <typeparam name="TTestClass">The test class. We use this class assembly to search for
+    /// assembly metadata attributes containing the content root paths to the web applications
+    /// with startup class <typeparamref name="TStartup"/>.</typeparam>
     /// <typeparam name="TStartup">The applications startup class.</typeparam>
-    public class WebApplicationTestFixture<TStartup> : IDisposable where TStartup : class
+    public class WebApplicationTestFixture<TTestClass, TStartup> : IDisposable where TTestClass : class where TStartup : class
     {
-        private readonly TestServer _server;
+        private TestServer _server;
+        private string _solutionRelativeContentRootPath;
+        private HttpClient _client;
+        private IWebHostBuilder _builder;
+        private readonly string _solutionSearchPattern;
 
         /// <summary>
         /// <para>
         /// Creates a TestServer instance using the MVC application defined by<typeparamref name="TStartup"/>.
         /// The startup code defined in <typeparamref name = "TStartup" /> will be executed to configure the application.
         /// </para>
         /// <para>
-        /// This constructor will infer the application root directive by searching for a solution file (*.sln) and then
-        /// appending the path<c>{AssemblyName}</c> to the solution directory.The application root directory will be
-        /// used to discover views and content files.
+        /// This constructor will infer the application root directive by searching for an <see cref="AssemblyMetadataAttribute"/>
+        /// on the <typeparamref name="TTestClass"/> assembly with the key 
+        /// Microsoft.AspNetCore.Testing.ContentRoot[<c><typeparamref name="TStartup" /> assembly file name</c>] and we will use
+        /// the value of that attribute as the content root. In case we can't find an attribute with the right key, we fall back to
+        /// searching for a solution file (*.sln) and then appending the path<c>{AssemblyName}</c> to the solution directory.
+        /// The application root directory will be used to discover views and content files.
```

Comment:
Is there another way to find the test assembly? Like does that AppContext or one of those statics have it?